psydac/feec/derivatives.py

```diff
@@ -26,51 +27,221 @@
+
+        # check that the number of points conforms
+        assert all([vn==wn+1 if not vwp and diffdir==i else vn==wn
+            for i, (vn, wn, vwp) in enumerate(zip(V.npts, W.npts, V.periods))])
+
+        self._spaceV = V
+        self._spaceW = W
+        self._diffdir = diffdir
+        self._negative = negative
+        self._transposed = transposed
+
+        if self._transposed:
+            self._domain = W
+            self._codomain = V
+        else:
+            self._domain = V
+            self._codomain = W
+        
+        # the local area in the codomain without padding
+        self._idslice = tuple([slice(pad, e-s+1+pad) for pad, s, e
+            in zip(self._codomain.pads, self._codomain.starts, self._codomain.ends)])
 
-    periodic : bool
-        True if matrix is periodic, False otherwise.
+        # prepare the slices (they are of the right size then, we checked this already)
+        # identity slice
+        idslice = self._idslice
+        
+        # differentiation slice (moved by one in the direction of differentiation)
+        diff_pad = self._codomain.pads[self._diffdir]
+        diff_s = self._codomain.starts[self._diffdir]
+        diff_e = self._codomain.ends[self._diffdir]
+
+        # the diffslice depends on the transposition
+        if self._transposed:
+            diff_partslice = slice(diff_pad-1, diff_e-diff_s+1+diff_pad-1)
+        else:
+            diff_partslice = slice(diff_pad+1, diff_e-diff_s+1+diff_pad+1)
+        
+        diffslice = tuple([diff_partslice if i==self._diffdir else idslice[i]
+                            for i in range(self._domain.ndim)])
+        
 
-    Results
-    -------
-    M : psydac.linalg.stencil.StencilMatrix
-        1D incidence matrix of shape (n, n) in the periodic case, and (n, n-1) otherwise.
+        # define differentiation lambda based on the parameter negative (or sign)
+        if self._negative:
+            self._do_diff = lambda v,out: np.subtract(v._data[idslice],
+                                v._data[diffslice], out=out._data[idslice])
+        else:
+            self._do_diff = lambda v,out: np.subtract(v._data[diffslice],
+                                v._data[idslice], out=out._data[idslice])
 
-    """
-    m = n if periodic else n - 1
+    @property
+    def domain(self):
+        return self._domain
+    
+    @property
+    def codomain(self):
+        return self._codomain
 
-    V1 = StencilVectorSpace([n], [p], [periodic])
-    V2 = StencilVectorSpace([m], [p], [periodic])
-    M  = StencilMatrix(V1, V2)
+    def dot(self, v, out=None):
+        """
+        Applies the differential operator on the given StencilVector.
 
-    for i in range(m):
-        M._data[p+i, p]   = -1.
-        M._data[p+i, p+1] =  1.
+        This operator will allocate temporary memory,
+        when used on a vector in-place (i.e. if `v is out`);
+        though this should not happen anyways due to `domain is not codomain`.
```

Comment:
Indeed, the case of `v is out` definitely cannot happen, as the two vectors must belong to different spaces. Maybe this was a note to self that ended up in the docstring?

Reply:
also, the derivative is strong here so there's no way you can map in the same spline space. This could happen with Fourier modes, but not splines (the ndofs could be the same, but the degrees will not)

Reply:
Yes, I agree; it won't happen in any of the `DiffOperator` subclasses, or with the `Derham` classes.

But, if somebody completely using the class out of place (since we export it, it's possible); as in this piece of code:

```
V = StencilVectorSpace([10], [2], [True])
v = V.zeros()
op = KroneckerDifferentialOperator(V, V, 0) # old name for this example
print(op.dot(v, out=v))
```

This code runs through without an error right now, and does the operation in-place (and hence allocates additional memory); since we work directly on the vector spaces, we have no association to a spline space; therefore this doesn't block us from using it this way. (that the code example I showed most likely therefore doesn't have any practical relevance)

I know, it's only very marginally important; but for the comment I thought it would be good to know when this operation might allocate additional memory (or not) through numpy.

Reply:
@davschneller OK, I understand, you're making a good argument. Please just make sure that the docstring is clear in this sense.